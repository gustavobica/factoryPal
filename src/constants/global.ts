export const HIGHLIGHT_COLOR = "#86baf1";

export enum Categories {
  efficiency = "efficiency",
  shift = "shift",
  downtime = "downtime",
}

export enum MetricTypes {
  percentage = "percentage",
  number = "number",
  secs = "secs",
  hours = "hours",
  mins = "mins",
}

export type TimeTypes = Extract<
  MetricTypes,
  MetricTypes.secs | MetricTypes.hours | MetricTypes.mins
>;

export const TimeTypesObject = {
  [MetricTypes.secs]: MetricTypes.secs,
  [MetricTypes.mins]: MetricTypes.mins,
  [MetricTypes.hours]: MetricTypes.hours,
} as const;

export type PercentageTypes = Extract<MetricTypes, MetricTypes.percentage>;

export const PercentageTypesObject = {
  [MetricTypes.percentage]: MetricTypes.percentage,
} as const;

export const DEFAULT_CATEGORY = "All";

export const SMALL_SCREEN = "550px";
