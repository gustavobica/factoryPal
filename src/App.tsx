import { useCallback, useMemo, useState } from "react";

import DashboardView from "./Views/DashboardView/DashboardView";
import DownTimeView from "./Views/DownTimeView";
import EfficiencyView from "./Views/EfficiencyView/EfficiencyView";
import ShiftView from "./Views/ShiftView/ShiftView";
import Logo from "./components/Logo/Logo";
import MetricTable from "./components/MetricsTable/MetricsTable";
import Selector from "./components/Selector/Selector";
import { capitalize } from "./components/utils/utils";
import { Categories, DEFAULT_CATEGORY } from "./constants/global";
import { StyledApp } from "./styles/global.styles";
import { Metric } from "./types/data";

const metricsData: Metric[] = require("./data.json").data;

function App() {
  // Metric that will be highlighted both on table and chart
  const [highlightedMetric, setHighlightedMetric] = useState("");
  const [selectedCategory, setSelectedCategory] = useState(DEFAULT_CATEGORY);

  const handleCategoryChange = useCallback((value: string) => {
    setSelectedCategory(value);
  }, []);

  // Filter by the category, if all selected use default data
  const filteredData = useMemo(
    () =>
      selectedCategory === DEFAULT_CATEGORY
        ? metricsData
        : metricsData.filter((metric) => metric.category === selectedCategory),
    [selectedCategory]
  );

  //Get a unique set of categories to use in the select (to switch views)
  const options = useMemo(
    () => Array.from(new Set(metricsData.map((metric) => metric.category))),
    []
  );

  /* If the category is unknow (not typed before) or All, return  give generic view.
   * If we know the category provide the correct view
   */

  return (
    <StyledApp>
      <Logo />
      <h1>Factory Metrics Dashboard - {capitalize(selectedCategory)}</h1>

      <Selector
        options={options}
        onChange={handleCategoryChange}
        selected={selectedCategory}
      />
      <div>
        {selectedCategory === Categories.efficiency && (
          <EfficiencyView
            metrics={filteredData}
            highlight={highlightedMetric}
            onHighlight={setHighlightedMetric}
          />
        )}
        {selectedCategory === Categories.shift && (
          <ShiftView
            metrics={filteredData}
            highlight={highlightedMetric}
            onHighlight={setHighlightedMetric}
          />
        )}
        {selectedCategory === Categories.downtime && (
          <DownTimeView
            metrics={filteredData}
            highlight={highlightedMetric}
            onHighlight={setHighlightedMetric}
          />
        )}

        {!(selectedCategory in Categories) && (
          <DashboardView
            metrics={filteredData}
            highlight={highlightedMetric}
            onHighlight={setHighlightedMetric}
          />
        )}
      </div>

      <MetricTable
        data={filteredData}
        highlight={highlightedMetric}
        onHighlight={setHighlightedMetric}
      />
    </StyledApp>
  );
}

export default App;
