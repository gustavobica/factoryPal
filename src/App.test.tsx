import { render, screen } from "@testing-library/react";

import userEvent from "@testing-library/user-event";
import App from "./App";
import { Categories } from "./constants/global";

describe("App component", () => {
  it("renders with default settings", () => {
    render(<App />);
    expect(
      screen.getByText(/Factory Metrics Dashboard - All/)
    ).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
  });

  it("allows navigation between categories", async () => {
    render(<App />);

    userEvent.selectOptions(
      screen.getByRole("combobox"),
      Categories.efficiency
    );
    expect(
      screen.getByText(/Factory Metrics Dashboard - Efficiency/i)
    ).toBeInTheDocument();

    userEvent.selectOptions(screen.getByRole("combobox"), Categories.downtime);
    expect(
      screen.getByText(/Factory Metrics Dashboard - Downtime/i)
    ).toBeInTheDocument();

    userEvent.selectOptions(screen.getByRole("combobox"), Categories.shift);
    expect(
      screen.getByText(/Factory Metrics Dashboard - Shift/i)
    ).toBeInTheDocument();
  });
  it("renders circular graphs for percentage metrics", () => {
    render(<App />);
    const percentageMetric = screen.getByText(/68%/); // Assume the circular graph displays percentage inside it

    expect(percentageMetric).toBeInTheDocument();
  });

  it("correctly highlights a metric", async () => {
    render(<App />);
    const firstMetric = screen.getByText(/68%/);
    userEvent.click(firstMetric);

    expect(firstMetric).toMatchSnapshot();
  });

  it("handles unknown metric types gracefully", () => {
    render(<App />);
    expect(
      screen.getByTestId(/MetricCard-cln_shift_years/)
    ).toBeInTheDocument();
  });
});
