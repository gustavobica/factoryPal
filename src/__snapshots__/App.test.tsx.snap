// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component correctly highlights a metric 1`] = `
<div
  class="sc-kuWfzm jEmpbQ"
>
  68
  %
</div>
`;
